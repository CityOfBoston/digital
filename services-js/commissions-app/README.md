### Deploys

- 2020.09.30: Security Patch: Remove hardcoded ssl pass in deploy script
- 2022.01.12: `Temp` Disable resume/cover letter upload and use mailto
