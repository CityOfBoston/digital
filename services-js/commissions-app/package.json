{
  "name": "commissions-app",
  "version": "0.0.0",
  "description": "Boards and Commissions GraphQL server and webapp",
  "private": true,
  "license": "CC0-1.0",
  "scripts": {
    "prebuild": "rimraf build && npm run generate-graphql-schema && npm run generate-graphql-types",
    "build": "concurrently \"npm run build:server\" \"npm run build:next\"",
    "build:server": "tsc -p tsconfig.server.json",
    "build:next": "next build src",
    "dev": "npm run prebuild && tsc-watch -p tsconfig.server.json --onSuccess \"npm run start\"",
    "storybook": "start-storybook -p 9001",
    "watch-dependencies": "lerna run --parallel --scope commissions-app --include-filtered-dependencies watch",
    "prepare": "npm run build",
    "start": "node build/server/start",
    "pretest": "tsc --noEmit",
    "test": "scripts/test.js",
    "generate-graphql-schema": "mkdir -p graphql && ts2gql src/server/graphql/schema.ts > graphql/schema.graphql",
    "generate-graphql-types": "apollo-codegen generate src/client/**/*.ts --project-name commissions-app --target typescript --output src/client/graphql/queries.d.ts && prettier --write src/client/graphql/queries.d.ts",
    "generate-fixtures": "ts-node ./scripts/generate-fixtures.ts",
    "generate-db-types": "ts-node ./scripts/generate-db-types.ts",
    "travis-deploy": "travis-service-deploy deploy/Dockerfile"
  },
  "jest": {
    "preset": "@cityofboston/config-jest-babel"
  },
  "resolutions": {
    "graphql": "0.13.2"
  },
  "dependencies": {
    "@babel/runtime": "7.0.0-beta.42",
    "@cityofboston/deploy-tools": "^0.0.0",
    "@cityofboston/hapi-common": "^0.0.0",
    "@cityofboston/hapi-next": "^0.0.0",
    "@cityofboston/mssql-common": "^0.0.0",
    "@cityofboston/next-client-common": "^0.0.0",
    "@cityofboston/percy-common": "^0.0.0",
    "@cityofboston/srv-decrypt-env": "^0.0.0",
    "@types/yup": "^0.24.7",
    "apollo-server-hapi": "^1.3.6",
    "boom": "^7.2.0",
    "dataloader": "^1.4.0",
    "dotenv": "^5.0.1",
    "formik": "^0.11.11",
    "graphql": "^0.13.2",
    "graphql-tools": "^3.0.0",
    "hapi": "^17.4.0",
    "hapi-accept-language2": "^2.0.3",
    "isomorphic-fetch": "^2.2.1",
    "mssql": "^4.2.1",
    "next": "6.1.1-canary.5",
    "node-cleanup": "^2.1.2",
    "react": "^16.4.0",
    "react-dom": "^16.4.0",
    "yup": "^0.25.1"
  },
  "devDependencies": {
    "@babel/cli": "7.0.0-beta.42",
    "@babel/core": "7.0.0-beta.42",
    "@cityofboston/config-babel": "^0.0.0",
    "@cityofboston/config-jest-babel": "^0.0.0",
    "@cityofboston/config-typescript": "^0.0.0",
    "@cityofboston/graphql-typescript": "^0.0.0",
    "@cityofboston/mssql-typescript": "^0.0.0",
    "@cityofboston/react-fleet": "^0.0.0",
    "@percy-io/percy-storybook": "^1.4.1",
    "@storybook/addon-actions": "^4.0.0-alpha.16",
    "@storybook/addon-storyshots": "^4.0.0-alpha.16",
    "@storybook/addons": "^4.0.0-alpha.16",
    "@storybook/react": "^4.0.0-alpha.16",
    "@types/isomorphic-fetch": "^0.0.34",
    "@types/jest": "^22.2.3",
    "@types/mssql": "^4.0.10",
    "@types/node": "^9.6.5",
    "@types/react": "^16.4.7",
    "@types/storybook__addon-actions": "^3.0.3",
    "@types/storybook__addon-storyshots": "^3.4.1",
    "@types/storybook__react": "^3.0.7",
    "@zeit/next-typescript": "^1.1.0",
    "apollo-codegen": "CityOfBoston/apollo-codegen#dist",
    "babel-core": "^7.0.0-0",
    "babel-plugin-require-context-hook": "^1.0.0",
    "concurrently": "^3.5.1",
    "eslint-plugin-graphql": "^2.1.1",
    "faker": "^4.1.0",
    "jest": "^22.4.0",
    "prettier": "^1.12.0",
    "raw-loader": "^0.5.1",
    "react-test-renderer": "^16.4.0",
    "rimraf": "^2.6.2",
    "ts-node": "^6.0.5",
    "ts2gql": "CityOfBoston/ts2gql#4f10fcce",
    "tsc-watch": "^1.0.26",
    "typescript": "^3.0.0"
  }
}
