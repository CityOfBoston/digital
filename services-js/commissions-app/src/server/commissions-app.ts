/* eslint no-console: 0 */

import fs from 'fs';

import { Server as HapiServer } from 'hapi';
import { boomify } from 'boom';
import Inert from 'inert';
import cleanup from 'node-cleanup';
import acceptLanguagePlugin from 'hapi-accept-language2';
const next = require('next');
import Rollbar from 'rollbar';
import { Client as PostmarkClient } from 'postmark';
import { ApolloServer } from 'apollo-server-hapi';

import {
  API_KEY_CONFIG_KEY,
  GRAPHQL_PATH_KEY,
  HAPI_INJECT_CONFIG_KEY,
  GOOGLE_TRACKING_ID_KEY,
} from '@cityofboston/next-client-common';

import {
  loggingPlugin,
  adminOkRoute,
  headerKeys,
  HeaderKeysOptions,
  rollbarPlugin,
  makeStaticAssetRoutes,
  HapiGraphqlContextFunction,
  rollbarErrorExtension,
} from '@cityofboston/hapi-common';

import { makeRoutesForNextApp } from '@cityofboston/hapi-next';

import { createConnectionPool } from '@cityofboston/mssql-common';
import decryptEnv from '@cityofboston/srv-decrypt-env';

import graphqlSchema, { Context } from './graphql/schema';
import CommissionsDao, { DbBoard } from './dao/CommissionsDao';
import CommissionsDaoFake from './dao/CommissionsDaoFake';
import { ConnectionPool } from 'mssql';
import { applyFormSchema, ApplyFormValues } from '../lib/validationSchema';

import Email from './services/Email';

const PATH_PREFIX = '/commissions';
const dev = !process.env.NODE_ENV || process.env.NODE_ENV === 'development';

export async function makeServer(port, rollbar: Rollbar) {
  let makeCommissionsDao: () => CommissionsDao;
  let commissionsDbPool: ConnectionPool | null = null;

  const postmarkClient = new PostmarkClient(
    process.env.POSTMARK_SERVER_API_TOKEN || 'fake-postmark-key'
  );

  const emailService = new Email(
    process.env.POSTMARK_FROM_ADDRESS || 'no-reply@boston.gov',
    process.env.POLICY_OFFICE_TO_ADDRESS || 'boardsandcommissions@boston.gov',
    process.env.COMMISSIONS_URI ||
      'http://zpappweb01/cityclerk/commissions/applications',
    postmarkClient,
    rollbar
  );

  if (
    process.env.ALLOW_FAKES ||
    process.env.NODE_ENV === 'test' ||
    (process.env.NODE_ENV !== 'production' &&
      !process.env.COMMISSIONS_DB_SERVER)
  ) {
    makeCommissionsDao = () => new CommissionsDaoFake() as any;
  } else {
    const username = process.env.COMMISSIONS_DB_USERNAME;
    const password = process.env.COMMISSIONS_DB_PASSWORD;
    const database = process.env.COMMISSIONS_DB_DATABASE;
    const domain = process.env.COMMISSIONS_DB_DOMAIN;
    const serverName = process.env.COMMISSIONS_DB_SERVER;

    if (!username) {
      throw new Error('Must specify COMMISSIONS_DB_USERNAME');
    }

    if (!password) {
      throw new Error('Must specify COMMISSIONS_DB_PASSWORD');
    }

    if (!database) {
      throw new Error('Must specify COMMISSIONS_DB_DATABASE');
    }

    if (!serverName) {
      throw new Error('Must specify COMMISSIONS_DB_SERVER');
    }

    commissionsDbPool = await createConnectionPool(
      {
        username,
        password,
        database,
        domain,
        server: serverName,
      },
      err => {
        console.error(err);
        rollbar.error(err);
        process.exit(-1);
      }
    );

    makeCommissionsDao = () => new CommissionsDao(commissionsDbPool!);
  }

  const serverOptions = {
    host: '0.0.0.0',
    port,
    tls: undefined as any,
    debug: dev
      ? {
          request: ['handler'],
        }
      : {},
  };

  if (process.env.USE_SSL) {
    serverOptions.tls = {
      key: fs.readFileSync('server.key'),
      cert: fs.readFileSync('server.crt'),
    };
  }

  const server = new HapiServer(serverOptions);

  // We don't turn on Next for test mode because it hangs Jest.
  let nextApp;

  if (process.env.NODE_ENV !== 'test') {
    // We load the config ourselves so that we can modify the runtime configs
    // from here.
    const config = require('../../next.config.js');

    config.publicRuntimeConfig = {
      ...config.publicRuntimeConfig,
      [GRAPHQL_PATH_KEY]: '/commissions/graphql',
      [API_KEY_CONFIG_KEY]: process.env.WEB_API_KEY,
      [GOOGLE_TRACKING_ID_KEY]: process.env.GOOGLE_TRACKING_ID,
    };

    config.serverRuntimeConfig = {
      ...config.serverRuntimeConfig,
      [HAPI_INJECT_CONFIG_KEY]: server.inject.bind(server),
    };

    nextApp = next({
      dev,
      dir: 'src',
      conf: config,
    });
  } else {
    nextApp = null;
  }

  server.auth.scheme('headerKeys', headerKeys);
  server.auth.strategy('apiHeaderKeys', 'headerKeys', {
    header: 'X-API-KEY',
    keys: process.env.API_KEYS ? process.env.API_KEYS.split(',') : [],
  } as HeaderKeysOptions);

  await server.register(acceptLanguagePlugin);
  await server.register({ plugin: rollbarPlugin, options: { rollbar } });
  await server.register(Inert);

  // We start up the server in test, and we don’t want it logging.
  if (process.env.NODE_ENV !== 'test') {
    await server.register(loggingPlugin);
  }

  server.route({
    method: 'GET',
    path: '/commissions',
    handler: (_, h) => h.redirect('/commissions/apply'),
  });

  server.route({
    method: 'POST',
    path: '/commissions/submit',
    options: {
      payload: {
        // Two 5mb files in bytes, plus another 500k for other form information.
        maxBytes: 1024 * 1024 * (2 * 5 + 0.5),
      },
    },
    handler: async (req, h) => {
      console.info('IN SUBMIT HANDLER');
      let validForm: ApplyFormValues;
      let fetchedBoards: DbBoard[];
      let applicationId: number;

      try {
        validForm = await applyFormSchema.validate(req.payload as any);
        fetchedBoards = await makeCommissionsDao().fetchBoards();
      } catch (e) {
        throw boomify(e, { statusCode: 400 });
      }

      applicationId = await makeCommissionsDao().apply(validForm);

      await emailService.sendConfirmations(
        validForm,
        fetchedBoards,
        applicationId
      );

      return h.response('ok');
    },
  });

  server.route(adminOkRoute);
  server.route(makeStaticAssetRoutes('/commissions/'));

  if (nextApp) {
    server.route(makeRoutesForNextApp(nextApp, '/commissions/'));
  }

  const context: HapiGraphqlContextFunction<Context> = () => ({
    commissionsDao: makeCommissionsDao(),
  });

  const apolloServer = new ApolloServer({
    context,
    schema: graphqlSchema,
    extensions: [rollbarErrorExtension(rollbar)],
  });

  await apolloServer.applyMiddleware({
    app: server,
    path: PATH_PREFIX + '/graphql',
    route: {
      auth: process.env.API_KEYS ? 'apiHeaderKeys' : false,
      cors: {
        origin: [
          'http://localhost:*',
          'https://localhost:*',
          'https://*.boston.gov',
          ...(process.env.CORS_ORIGINS
            ? process.env.CORS_ORIGINS.split(',')
            : []),
        ],
        additionalHeaders: ['X-API-KEY'],
      },
    },
  });

  return {
    server,
    startup: async () => {
      await Promise.all([server.start(), nextApp ? nextApp.prepare() : null]);

      console.log(
        `> Ready on http${
          process.env.USE_SSL ? 's' : ''
        }://localhost:${port}${PATH_PREFIX}`
      );

      // Add more shutdown code here.
      return () =>
        Promise.all([
          server.stop(),
          commissionsDbPool ? commissionsDbPool.close() : Promise.resolve(),
        ]);
    },
  };
}

export default async function startServer(rollbar: Rollbar) {
  await decryptEnv();

  const port = parseInt(process.env.PORT || '3000', 10);

  const { startup } = await makeServer(port, rollbar);
  const shutdown = await startup();

  // tsc-watch sends SIGUSR2 when it’s time to restart. That’s not caught by
  // cleanup, so we get it ourselves so we can do a clean shutdown.
  process.on('SIGUSR2', () => {
    // Keeps us alive
    process.stdin.resume();

    // This will cause cleanup to run below
    process.kill(process.pid, 'SIGINT');
  });

  cleanup(exitCode => {
    shutdown().then(
      () => {
        process.exit(exitCode);
      },
      err => {
        rollbar.error(err);
        console.log('CLEAN EXIT FAILED', err);
        process.exit(-1);
      }
    );

    cleanup.uninstall();
    return false;
  });
}
