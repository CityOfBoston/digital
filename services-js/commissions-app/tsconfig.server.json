{
  "extends": "./node_modules/@cityofboston/config-typescript/tsconfig.default.json",
  "compilerOptions": {
    "rootDir": "src",
    "outDir": "build",
    "declaration": false,
    "declarationMap": false
  },
  "include": [
    "src/**/*"
  ],
  "exclude": [
    "src/client/**/*",
    "src/pages/**/*",
    "src/stories/**/*"
  ]
}