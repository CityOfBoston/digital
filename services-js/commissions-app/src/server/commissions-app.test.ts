/* eslint no-console: 0 */
import { makeServer } from './commissions-app';
import { Server } from 'hapi';

// eslint-disable-next-line no-undef
jasmine.DEFAULT_TIMEOUT_INTERVAL = 15000;

const PATH_PREFIX = '/commissions';
const API_KEY = 'test-api-key';
const GRAPHQL_INTROSPECTION_PAYLOAD = {
  query: `
    {
      __schema {
        types {
          name
        }
      }
    }`,
};

describe('server creation', () => {
  it('can be created', async () => {
    const { server } = await makeServer(0, null as any);
    expect(server).toBeDefined();
  });
});

describe('server', () => {
  let server: Server;
  let shutdown;

  beforeAll(async () => {
    const oldLog = console.log;
    console.log = () => {};

    process.env.API_KEYS = API_KEY;

    const out = await makeServer(0, null as any);
    server = out.server;
    const startup = out.startup;

    shutdown = await startup();
    console.log = oldLog;
  });

  afterAll(async () => {
    await shutdown();
  });

  it('passes healthcheck', async () => {
    const resp = await server.inject({
      method: 'GET',
      url: '/admin/ok',
    });

    expect(resp.statusCode).toEqual(200);
  });

  it('handles GraphQL query', async () => {
    const resp = await server.inject({
      method: 'POST',
      url: `${PATH_PREFIX}/graphql`,
      headers: {
        'X-API-KEY': API_KEY,
      },
      payload: GRAPHQL_INTROSPECTION_PAYLOAD,
    });

    expect(resp.statusCode).toEqual(200);
    expect(JSON.parse(resp.payload)).toMatchSnapshot();
  });
});
