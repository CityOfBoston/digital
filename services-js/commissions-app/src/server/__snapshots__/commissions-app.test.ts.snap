// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`server handles GraphQL query 1`] = `
Object {
  "data": Object {
    "__schema": Object {
      "types": Array [
        Object {
          "name": "Query",
        },
        Object {
          "name": "Float",
        },
        Object {
          "name": "Commission",
        },
        Object {
          "name": "String",
        },
        Object {
          "name": "Department",
        },
        Object {
          "name": "Member",
        },
        Object {
          "name": "Date",
        },
        Object {
          "name": "Boolean",
        },
        Object {
          "name": "PolicyType",
        },
        Object {
          "name": "__Schema",
        },
        Object {
          "name": "__Type",
        },
        Object {
          "name": "__TypeKind",
        },
        Object {
          "name": "__Field",
        },
        Object {
          "name": "__InputValue",
        },
        Object {
          "name": "__EnumValue",
        },
        Object {
          "name": "__Directive",
        },
        Object {
          "name": "__DirectiveLocation",
        },
      ],
    },
  },
}
`;
